cassandane/Cassandane/Cyrus/SearchFuzzy.pm

```diff
@@ -1926,5 +1920,306 @@ sub test_reindex_mb_uniqueid
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert extractor only got called once";
+    my @tracefiles = glob($tracedir."/*_PUT_*");
+    $self->assert_num_equals(1, scalar @tracefiles);
+
+    xlog "Assert cache contains one file";
+    my @files = glob($cachedir."/*");
+    $self->assert_num_equals(1, scalar @files);
+}
+
+sub test_squatter_attachextract_cachedir_noperm
+    :min_version_3_9 :needs_search_xapian :SearchAttachmentExtractor :NoCheckSyslog
+{
+    my ($self) = @_;
+    my $instance = $self->{instance};
+    my $imap = $self->{store}->get_client();
+
+    my $tracedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    $self->start_echo_extractor(tracedir => $tracedir);
+
+    xlog "Run squatter with read-only cache directory";
+    my $cachedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    chmod "400", $cachedir;
+    $self->squatter_attachextract_cache_run($cachedir, "--allow-partials");
+
+    xlog "Assert text bodies of both messages are indexed";
+    my $uids = $imap->search('fuzzy', 'body', 'bodyterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert attachments of both messages are not indexed";
+    $uids = $imap->search('fuzzy', 'xattachmentbody', 'attachterm');
+    $self->assert_deep_equals([], $uids);
+
+    xlog "Assert extractor got called twice with attachment uploads";
+    my @tracefiles = glob($tracedir."/*_PUT_*");
+    $self->assert_num_equals(2, scalar @tracefiles);
+
+    xlog "Assert cache contains no file";
+    chmod "0700", $cachedir;
+    my @files = glob($cachedir."/*");
+    $self->assert_num_equals(0, scalar @files);
+}
+
+sub test_squatter_attachextract_cacheonly
+    :min_version_3_9 :needs_search_xapian :SearchAttachmentExtractor :NoCheckSyslog
+{
+    my ($self) = @_;
+    my $instance = $self->{instance};
+    my $imap = $self->{store}->get_client();
+
+    my $tracedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    $self->start_echo_extractor(tracedir => $tracedir);
+
+    xlog "Instruct squatter to only use attachextract cache";
+    my $cachedir = tempdir(DIR => $instance->{basedir} . "/tmp");
+    $self->squatter_attachextract_cache_run($cachedir,
+        "--attachextract-cache-only", "--allow-partials");
+
+    xlog "Assert text bodies of both messages are indexed";
+    my $uids = $imap->search('fuzzy', 'body', 'bodyterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert attachments of both messages are not indexed";
+    $uids = $imap->search('fuzzy', 'xattachmentbody', 'attachterm');
+    $self->assert_deep_equals([], $uids);
+
+    xlog "Assert extractor did not get got called";
+    my @tracefiles = glob($tracedir."/*");
+    $self->assert_num_equals(0, scalar @tracefiles);
+
+    xlog "Assert cache contains no file";
+    my @files = glob($cachedir."/*");
+    $self->assert_num_equals(0, scalar @files);
+}
+
+sub test_squatter_attachextract_nolock_slow
```

Comment:
This test failed for me locally once, but now that I'm trying to reproduce it to find out why, it's consistently passing.  Curious.

Reply:
For that matter, it only takes six wallclock seconds to run, so you can probably remove the `_slow` and just let it run every time.  Likewise for `..._timeout_slow`, it only takes five wallclock seconds.

Reply:
Removed `_slow` for both tests